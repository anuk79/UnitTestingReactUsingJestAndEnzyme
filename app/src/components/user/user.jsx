// @flow

// vendors
import React from 'react';
import { debounce } from 'throttle-debounce';

const UserView = (props) => {
    const testFunction = () => {
        alert('testing debounce');
    }
    debounce(300, testFunction);
    return (
        <React.Fragment>
            <h3>{'User Details:'}</h3>
            <div>
                <label>{'First Name: '}</label>
                <span>{props.userDetails.firstName}</span>
            </div>
            <div>
                <label>{'Last Name: '}</label>
                <span>{props.userDetails.lastName}</span>
            </div>
        </React.Fragment>
    );
}

export default UserView;