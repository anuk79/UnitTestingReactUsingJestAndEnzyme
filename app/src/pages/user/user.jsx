// @flow

// vendors
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

// components
import UserView from '../../components/user/user';
import LoadingSpinner from '../../components/loading-spinner/loading-spinner';

// actions
import { fetchUserDetails } from '../../actions/user.action';

const mapStateToProps = state => ({
    userDetails: state.user.userDetails,
    fetching: state.user.fetching,
    errorFlag: state.user.errorFlag
});

const mapDispatchToProps = dispatch => ({
    fetchUserDetails: bindActionCreators(fetchUserDetails, dispatch)
});

export class User extends Component {
    componentDidMount() {
        this.props.fetchUserDetails();
    }

    render() {
        const { fetching, userDetails, errorFlag } = this.props;
        let content;
        if(fetching) {
            content = <LoadingSpinner customClassName="user-loading-spinner" />
        } else if(errorFlag) {
            content = (
                <div className="user-error">{'Some error occurred. Please try again later.'}</div>
            );
        } else {
            content = <UserView userDetails={userDetails} />
        }
        return content;
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(User);