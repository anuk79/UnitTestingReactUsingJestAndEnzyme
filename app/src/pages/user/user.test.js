
// vendors
import React from "react";
import { shallow } from 'enzyme';
import { shallowToJson } from "enzyme-to-json";

// components
import { User } from './user';

describe("User component", () => {
    it("should render correctly when data is fetching", () => {
        const mockFetchUserDetails = jest.fn();
        const wrapper = shallow(<User fetching={true} fetchUserDetails={mockFetchUserDetails}
            />);
        expect(shallowToJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('LoadingSpinner').length).toBe(1);
    });

    it("should render correctly when some error", () => {
        const mockFetchUserDetails = jest.fn();
        const wrapper = shallow(<User errorFlag={true} fetchUserDetails={mockFetchUserDetails}
            />);
        expect(shallowToJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('.user-error').length).toBe(1);
    });

    it("should render correctly when no error and data fetching is done", () => {
        const mockFetchUserDetails = jest.fn();
        const wrapper = shallow(<User userDetails={{}} fetchUserDetails={mockFetchUserDetails}
            />);
        expect(shallowToJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('UserView').length).toBe(1);
        expect(wrapper.find('UserView').prop('userDetails')).toEqual({});
        expect(mockFetchUserDetails).toBeCalled();
    });
});