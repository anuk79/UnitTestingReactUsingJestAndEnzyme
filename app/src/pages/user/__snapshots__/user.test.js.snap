// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component should render correctly when data is fetching 1`] = `
<LoadingSpinner
  customClassName="user-loading-spinner"
/>
`;

exports[`User component should render correctly when no error and data fetching is done 1`] = `
<UserView
  userDetails={Object {}}
/>
`;

exports[`User component should render correctly when some error 1`] = `
<div
  className="user-error"
>
  Some error occurred. Please try again later.
</div>
`;
